templates/routing-ortools/main.py

```diff
@@ -0,0 +1,177 @@
+"""
+
+import argparse
+import json
+import sys
+import time
+from typing import Any
+
+from ortools.constraint_solver import pywrapcp, routing_enums_pb2
+
+
+def main():
+    """Entry point for the template."""
+    parser = argparse.ArgumentParser(description="Solve problems with OR-Tools.")
+    parser.add_argument(
+        "-input",
+        default="",
+        help="Path to input file. Default is stdin.",
+    )
+    parser.add_argument(
+        "-output",
+        default="",
+        help="Path to output file. Default is stdout.",
+    )
+    parser.add_argument(
+        "-duration",
+        default=30,
+        help="Max runtime duration (in seconds). Default is 30.",
+        type=int,
+    )
+    args = parser.parse_args()
+
+    # Read input data, solve the problem and write the solution.
+    input_data = read_input(args.input)
+    log("Solving routing problem:")
+    log(f"  - num_vehicles: {input_data.get('num_vehicles', 0)}")
+    log(f"  - stops: {len(input_data.get('distance_matrix', []))-1}")
+    log(f"  - max duration: {args.duration} seconds")
+    solution = solve(input_data, args.duration)
+    write_output(args.output, solution)
+
+
+def solve(input_data: dict[str, Any], duration: int) -> dict[str, Any]:
+    """Solves the given problem and returns the solution."""
+
+    # Create the routing index manager.
+    manager = pywrapcp.RoutingIndexManager(
+        len(input_data["distance_matrix"]),
+        input_data["num_vehicles"],
+        input_data["depot"],
+    )
+
+    # Create Routing Model.
+    routing = pywrapcp.RoutingModel(manager)
+
+    # Create and register a transit callback.
+    def distance_callback(from_index, to_index):
+        """Returns the distance between the two nodes."""
+        # Convert from routing variable Index to distance matrix NodeIndex.
+        from_node = manager.IndexToNode(from_index)
+        to_node = manager.IndexToNode(to_index)
+        return input_data["distance_matrix"][from_node][to_node]
+
+    transit_callback_index = routing.RegisterTransitCallback(distance_callback)
+
+    # Define cost of each arc.
+    routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)
+
+    # Add Distance constraint.
+    dimension_name = "Distance"
+    routing.AddDimension(
+        transit_callback_index,
+        0,  # no slack
+        3000,  # vehicle maximum travel distance
+        True,  # start cumul to zero
+        dimension_name,
+    )
+    distance_dimension = routing.GetDimensionOrDie(dimension_name)
+    distance_dimension.SetGlobalSpanCostCoefficient(100)
+
+    # Setting first solution heuristic.
+    search_parameters = pywrapcp.DefaultRoutingSearchParameters()
+    search_parameters.first_solution_strategy = routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC
+    search_parameters.time_limit.FromSeconds(duration)
+
+    # Solve the problem.
+    start_time = time.time()
+    solution = routing.SolveWithParameters(search_parameters)
```

Comment:
We need to handle the case where no solution is found. E.g., the solver ran out of time for a highly constrained input or the input is infeasible. In both cases the returned solution will simply be `None`, I believe.